core/chains/evm/types/models.go

```diff
@@ -90,10 +94,49 @@ func (h *Head) BlockDifficulty() *big.Int {
 
 // EarliestInChain recurses through parents until it finds the earliest one
 func (h *Head) EarliestInChain() *Head {
-	for h.Parent != nil {
-		h = h.Parent
+	var earliestInChain *Head
+	h.forEach(func(h *Head) bool {
+		earliestInChain = h
+		return true
+	})
+	return earliestInChain
+}
+
+// forEach - go through the chain linked list. Stops when do returns false or h is nil
+func (h *Head) forEach(do func(h *Head) bool) {
```

Comment:
We could use the new for-range `iter.Seq` API here. Your pattern is nearly identical:
```suggestion
// forEach - go through the chain linked list. Stops when do returns false or h is nil
func (h *Head) all() iter.Seq[*Head] {
	return func(yield func(*Head) bool) {
		...
		if !yield(cur) {
			return
		}
```
Then the callers can do regular for loops:
```diff
- h.forEach(func(h *Head) bool {
+ for _ := range h.all() {
```

Reply:
https://github.com/smartcontractkit/chainlink/pull/14353

Reply:
The iterator is no longer needed, as I've dropped the cycle detection (the reason is in the comment below).